### PR TITLE
Add support for KIP-110 ZSTD compression

### PR DESCRIPTION
**Pull request by vavrusa**
This was introduced recently in https://cwiki.apache.org/confluence/display/KAFKA/KIP-110%3A+Add+Codec+for+ZStandard+Compression

And merged in https://github.com/apache/kafka/pull/2267

I'm going to need some help figuring out how to test this. I think I added the prerequisites right - Fetch APIv10 and Produce APIv7.

### CODE REVIEW

File: src/rdkafka_msgset_writer.c [hunk 1]
```diff
@@ -948,7 +952,22 @@ rd_kafka_msgset_writer_compress_lz4 (rd_kafka_msgset_writer_t *msetw,
...
-
+#if WITH_ZSTD
+/**
+ * @brief Compress messageset using ZSTD
+ */
+static int
+rd_kafka_msgset_writer_compress_zstd (rd_kafka_msgset_writer_t *msetw,
+                                     rd_slice_t *slice, struct iovec *ciov) {
+        rd_kafka_resp_err_t err;
+        int comp_level =
```
Comment (edenhill) [inline src/rdkafka_msgset_writer.c]: micro nit: this temporary variable does not serve much purpose.

File: src/rdzstd.c [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+        rd_kafka_resp_err_t err = RD_KAFKA_RESP_ERR_NO_ERROR;
+        char *decompressed = NULL;
+
+        size_t out_bufsize = ZSTD_getFrameContentSize(inbuf, inlen);
+        switch (out_bufsize) {
+        case ZSTD_CONTENTSIZE_UNKNOWN:
+                /* Decompressed size cannot be determined, make a guess */
+                out_bufsize = inlen * 2;
+        case ZSTD_CONTENTSIZE_ERROR:
```
Comment (edenhill) [inline src/rdzstd.c]: should be a break; here, right?

File: src/rdzstd.c [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                if (!ZSTD_isError(ret)) {
+                        *outlenp = ret;
+                        *outbuf = decompressed;
+                        goto done;
+                }
+
+                /* Check if the destination size is too small */
+                ZSTD_ErrorCode err = ZSTD_getErrorCode(ret);
+                if (err == ZSTD_error_dstSize_tooSmall) {
+                        out_bufsize *= 2;
```
Comment (edenhill) [inline src/rdzstd.c]: I can't find the source now, and I don't really remember why, but it is supposedly better to multiply by 1.75 than 2 when increasing buffers.

https://github.com/edenhill/librdkafka/blob/master/src/rdkafka_lz4.c#L268

File: src/rdzstd.c [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+                        char *newp = rd_realloc(decompressed, out_bufsize);
```
Comment (edenhill) [inline src/rdzstd.c]: variable definitions need to go at the top of the scope to be compatible with older non C99 compilers (MSVC).

File: src/rdzstd.c (hunk 3 above)
Comment (edenhill) [reply src/rdzstd.c]: Add this: `rd_atomic64_add(&rkb->rkb_c.zbuf_grow, 1);`

File: src/rdzstd.c [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                /* Attempt to decompress */
+                size_t ret = ZSTD_decompress(decompressed, out_bufsize, inbuf, inlen);
+                if (!ZSTD_isError(ret)) {
+                        *outlenp = ret;
+                        *outbuf = decompressed;
+                        goto done;
+                }
+
+                /* Check if the destination size is too small */
+                ZSTD_ErrorCode err = ZSTD_getErrorCode(ret);
```
Comment (edenhill) [inline src/rdzstd.c]: variable definititions need to get at the top of the scope.

`err` shadows the top-level `rd_kafka_resp_err_t err`, better call this one `zerr`

File: src/rdzstd.c [hunk 6]
```diff
@@ -0,0 +1,197 @@
+/*
```
Comment (edenhill) [inline src/rdzstd.c]: Rename this file to rdkafka_zstd.c

File: src/rdzstd.c [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+ * Copyright (c) 2018, Marek Vavrusa <mvavrusa@cloudflare.com
```
Comment (edenhill) [inline src/rdzstd.c]: Missing trailing `>`.

You will need to Copyright it to Magnus Edenhill too.

File: src/rdzstd.c [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+                rd_free(decompressed);
+        }
+
+        return err;
+}
+
+rd_kafka_resp_err_t
+rd_kafka_zstd_compress (rd_kafka_broker_t *rkb, int comp_level,
+                       rd_slice_t *slice, void **outbuf, size_t *outlenp) {
+        ZSTD_CStream* cctx;
```
Comment (edenhill) [inline src/rdzstd.c]: `*` goes on the variable side, not type.

File: src/rdzstd.c [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+                                   "%s",
+                                   in.size, out.size - out.pos,
+                                   ZSTD_getErrorName(r));
+                        err = RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+                        goto done;
+                }
+
+                /* No space left in output buffer, but input isn't fully consumed */
+                if (in.pos < in.size) {
+                        err = RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
```
Comment (edenhill) [inline src/rdzstd.c]: Add a rd_rkb_dbg() saying this.

File: src/rdzstd.c [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+                        goto done;
+                }
+        }
+
+        rd_assert(rd_slice_remains(slice) == 0);
```
Comment (edenhill) [inline src/rdzstd.c]: Better to error out here instead of crashing since the input data is from an external source.

File: src/rdzstd.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2018, Marek Vavrusa <mvavrusa@cloudflare.com
```
Comment (edenhill) [inline src/rdzstd.h]: dito, rename to rdkafka_zstd.h, add copyright, add trailing `>`

File: CONFIGURATION.md [hunk 12]
```diff
@@ -140,3 +140,4 @@ offset.store.method                      |  C  | file, broker    |        broker
 consume.callback.max.messages            |  C  | 0 .. 1000000    |             0 | Maximum number of messages to dispatch in one `rd_kafka_consume_callback*()` call (0 = unlimited) <br>*Type: integer*
 
 ### C/P legend: C = Consumer, P = Producer, * = both
+
```
Comment (edenhill) [inline CONFIGURATION.md]: This file is automatically generated but seems to be missing the zstd stuff; make sure you build the source with `make` at least once to let auto-generation kick in.

File: LICENSE.zstd [hunk 13]
```diff
@@ -0,0 +1,30 @@
+BSD License
```
Comment (edenhill) [inline LICENSE.zstd]: Since we're not including the zstd source we shouldn't include this license.

File: LICENSES.txt [hunk 14]
```diff
@@ -342,3 +342,37 @@ For the files wingetopt.c wingetopt.h downloaded from https://github.com/alex85k
  */
 
 
+LICENSE.zstd
```
Comment (edenhill) [inline LICENSES.txt]: dito

**Comment by edenhill** [comment]
This will go in v1.0.0, which is currently the `idempotence` branch.
It would be great if you could retarget and rebase this on `idempotence`.

File: src/rdkafka_msgset_writer.c (hunk 1 above)
Comment (vavrusa) [reply src/rdkafka_msgset_writer.c]: it's the same as above, I think it's a bit neater as the rvalue is really long, but will remove it if you prefer that

File: src/rdzstd.c (hunk 3 above)
Comment (vavrusa) [reply src/rdzstd.c]: Updated

File: src/rdzstd.c (hunk 4 above)
Comment (vavrusa) [reply src/rdzstd.c]: I updated it but it doesn't compile with c89 (no inline etc.), is this is just for MSVC compatibility?

Comment (edenhill) [reply src/rdzstd.c]: Yeah, at least that's the most common weird old compiler we come across frequently, and it is stuck somewhere between c89 and c99 :)

File: src/rdkafka_msgset_writer.c (hunk 1 above)
Comment (edenhill) [reply src/rdkafka_msgset_writer.c]: you can leave it 👍

**Comment by vavrusa** [comment]
Yeah, we've been using zstd in Kafka for a while at Cloudflare, but it only currently "works" with Sarama, so I can't move without it.

**Comment by edenhill** [comment]
> .. so I can't move without it.

Move to what? :)

v1.0.0 is scheduled for release within a month.

**Comment by vavrusa** [comment]
> Move to what? :)

I wrote the Kafka table engine for ClickHouse using librdkafka, but can't use it with most topics without zstd. The librdkafka is embedded as a submodule there, so I don't have to wait for the release tag before testing it.

**Comment by vavrusa** [comment]
Rebased to `idempotence` branch.

**Comment by vavrusa** [comment]
Thanks for reviewing! 🙌

**Comment by Donis-** [comment]
@vavrusa did you get it to work in your case?

We upgraded to confluent kafka 5.1.0 / 2.1.0 and set the broker default compression to ZSTD.
Java clients seem to be working fine, but i can't get kafkacat with librdkafka to work:
Consumption fails. Based on logs, zstd was enabled via compilation and is supported by broker:

%7|1546470506.414|APIVERSION|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/bootstrap]: sasl_plaintext://xxx/bootstrap:  Feature ZSTD: Produce (7..7) supported by broker
%7|1546470506.415|APIVERSION|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/bootstrap]: sasl_plaintext://xxx/bootstrap:  Feature ZSTD: Fetch (10..10) supported by broker
%7|1546470506.417|APIVERSION|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/bootstrap]: sasl_plaintext://xxx/bootstrap: Enabling feature ZSTD
%7|1546470506.417|FEATURE|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/bootstrap]: sasl_plaintext://xxx/bootstrap: Updated enabled protocol features to MsgVer1,ApiVersion,BrokerBalancedConsumer,ThrottleTime,Sasl,SaslHandshake,BrokerGroupCoordinator,LZ4,OffsetTime,MsgVer2,IdempotentProducer,ZSTD
...
%7|1546470508.106|FETCH|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/105]: sasl_plaintext://xxx/105: Fetch topic XXX  [55] at offset 0 (v2)
%7|1546470508.108|FETCH|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/105]: sasl_plaintext://xxx/105: Fetch 1/1/16 toppar(s)
%7|1546470508.108|SEND|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/105]: sasl_plaintext://xxx/105: Sent FetchRequest (v4, 112 bytes @ 0, CorrId 3)
%7|1546470508.155|RECV|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/105]: sasl_plaintext://xxx/105: Received FetchResponse (v4, 92 bytes, CorrId 3, rtt 46.73ms)
%7|1546470508.156|FETCH|rdkafka#consumer-1| [thrd:sasl_plaintext://xxx/105]: sasl_plaintext://xxx/105: Topic XXX [55] MessageSet size 0, error "Err-76?", MaxOffset -1, Ver 2/2

76 is the error introduced with KIP-110: UNSUPPORTED_COMPRESSION_TYPE.
https://cwiki.apache.org/confluence/display/KAFKA/KIP-110%3A+Add+Codec+for+ZStandard+Compression

Maybe we need to bump the v4 fetch version somehow as per "Zstd will only be allowed for the bumped fetch API ".? or am i missing something?

**Comment by vavrusa** [comment]
Yes, it's been working for some time. Maybe the negotiation isn't working properly for your case? I don't know that much about Kafka itself to help you, sorry!

**Comment by Donis-** [comment]
thanks for confirming!
just to double check, you're both producing and consuming via librdkafka? 
@edenhill maybe you have an idea on what could be going wrong in my case?

**Comment by Donis-** [comment]
seems fetch api v10 support was missing, by adding the the fetch protocol support for v10 i got it to work. i can clean up and submit a PR.

**Pull request merged**